pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -96,63 +109,137 @@ func (e *DockerRegistryServiceController) Stop() {
 	}
 }
 
-func (e *DockerRegistryServiceController) getServiceLocation(service *api.Service) string {
+// waitForSync polls to see if all the controllers are synced.  If so, it then makes a call to
+// update the dockercfg secrets initially
+func (e *DockerRegistryServiceController) waitForSync(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-time.After(100 * time.Millisecond):
+			if e.servicesSynced() {
+				e.markForChange()
+				return
+			}
+		}
+	}
+
+}
+
+// markForChange adds a item to the work channel, but only if there's not already something there
+// Clayton says that len() has the expected happens-before semantic.  This effectively de-dups changes
+func (e *DockerRegistryServiceController) markForChange() {
+	e.workLock.Lock()
+	defer e.workLock.Unlock()
+
+	if len(e.work) == 0 {
+		e.work <- true
+	}
+}
+
+// watchForChanges reacts to queued attempts to handle
+func (e *DockerRegistryServiceController) watchForChanges(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-e.work:
+			// if we're notified, do work and then start waiting again.
+			e.handleLocationChange()
+		}
+	}
+
+}
+
+// getServiceLocations returns the dns form and the ip form of the sercie
+func (e *DockerRegistryServiceController) getServiceLocations(service *api.Service) (string, string) {
 	hasPortalIP := (len(service.Spec.ClusterIP) > 0) && (net.ParseIP(service.Spec.ClusterIP) != nil)
 	if hasPortalIP && len(service.Spec.Ports) > 0 {
-		return net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port))
+		return net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+			net.JoinHostPort(fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port))
 	}
 
-	return e.defaultDockerURL
+	// returning the same one twice will result in a series of stomp
+	return e.defaultDockerURL, e.defaultDockerURL
+}
+
+func (e *DockerRegistryServiceController) getServiceCacheKey() string {
```

Comment:
can we use NewIndexerInformer to avoid magically knowing the internal index key format?